src/DynamicExpresso.Core/Parsing/Parser.cs

```diff
@@ -1207,41 +1207,126 @@ private Expression[] ParseArrayInitializerList()
+			if (bindingList.Count == 0)
+			{
+				actions.Add(instance);
+				return Expression.Block(new ParameterExpression[] { instance }, actions);
+			}
+			return Expression.MemberInit(newExpr, bindingList.ToArray());
+		}
 
-				var propertyOrFieldName = _token.text;
-				var member = FindPropertyOrField(newType, propertyOrFieldName, false);
-				if (member == null)
-					throw CreateParseException(_token.pos, ErrorMessages.UnknownPropertyOrField, propertyOrFieldName, GetTypeName(newType));
+		private void ParsePossibleMemberBinding(NewExpression newExpr, Type newType, int originalPos, List<MemberBinding> bindingList, List<Expression> actions, ParameterExpression instance, bool allowCollectionInit)
+		{
+			if (actions.Count > 1)
+			{
+				if (
+						this._arguments.TryGetParameters(_token.text, out _)
+							||
+						this._arguments.TryGetIdentifier(_token.text, out _)
+							||
+						FindPropertyOrField(newType, _token.text, false) == null)
+				{
+					ParseCollectionInitalizer(newExpr, newType, originalPos, bindingList, actions, instance, allowCollectionInit);
+					return;
+				}
+				else
+				{
+					throw CreateParseException(originalPos, ErrorMessages.InvalidInitializerMemberDeclarator);
+				}
+			}
+			ValidateToken(TokenId.Identifier, ErrorMessages.IdentifierExpected);
 
-				NextToken();
+			var propertyOrFieldName = _token.text;
+			var member = FindPropertyOrField(newType, propertyOrFieldName, false);
+			if (member == null)
+			{
+				var pos = _token.pos;
+				if (allowCollectionInit)
+				{
+					NextToken();
+					if (_token.id != TokenId.Equal || _arguments.TryGetIdentifier(propertyOrFieldName, out _) || _arguments.TryGetParameters(propertyOrFieldName, out _))
+					{
+						SetTextPos(pos);
+						NextToken();
+						actions.Add(ParseNormalMethodInvocation(newType, instance, _token.pos, "Add", new[] { ParseExpressionSegment() }));
+						return;
+					}
+				}
+				throw CreateParseException(pos, ErrorMessages.UnknownPropertyOrField, propertyOrFieldName, GetTypeName(newType));
 
-				ValidateToken(TokenId.Equal, ErrorMessages.EqualExpected);
-				NextToken();
+			}
+			NextToken();
 
-				var value = ParseExpressionSegment();
-				bindingList.Add(Expression.Bind(member, value));
+			ValidateToken(TokenId.Equal, ErrorMessages.EqualExpected);
+			NextToken();
 
-				if (_token.id != TokenId.Comma) break;
+			var value = ParseExpressionSegment();
+			bindingList.Add(Expression.Bind(member, value));
+		}
+
+		private void ParseCollectionInitalizer(NewExpression newExpr, Type newType, int originalPos, List<MemberBinding> bindingList, List<Expression> actions, ParameterExpression instance, bool allowCollectionInit)
+		{
+			if (!allowCollectionInit)
+			{
+				throw CreateParseException(_token.pos, ErrorMessages.CollectionInitializationNotSupported, newType, typeof(IEnumerable));
+			}
+			
+			if (bindingList.Count > 0)
+			{
+				throw CreateParseException(originalPos, ErrorMessages.InvalidInitializerMemberDeclarator);
```

Comment:
Please add a unit test for this.

Reply:
covered in multiple tests under Ctor_InvalidInitializerMemberDeclarator test